File: python-backend/src/api/controllers/auth_controller.py
"""
auth_controller.py

Defines the API endpoints for user authentication, including login, registration,
and token refreshing. Utilizes dependency-injected authentication service.
Provides detailed exception handling and HTTP status code mapping.
"""

from fastapi import APIRouter, Depends, HTTPException
from src.application.interfaces.interfaces import IAuthService
from src.application.dtos.auth_dtos import LoginRequest, LoginResponse, RegisterRequest, RefreshTokenRequest
from src.api.dependencies import get_auth_service

# Create an API router for authentication-related routes
router = APIRouter(prefix="/auth", tags=["Auth"])

@router.post("/login", response_model=LoginResponse)
def login(request: LoginRequest, auth_service: IAuthService = Depends(get_auth_service)):
    """
    Authenticate a user and return a JWT access/refresh token pair.

    Args:
        request (LoginRequest): The login credentials from the client.
        auth_service (IAuthService): Auth service for performing authentication.

    Returns:
        LoginResponse: JWT tokens and user info on success.

    Raises:
        HTTPException: 401 if credentials are invalid.
    """
    try:
        # Attempt to authenticate the user using provided credentials
        return auth_service.login(request)
    except Exception as e:
        # Authentication failed (invalid user, wrong password, etc.)
        raise HTTPException(status_code=401, detail=str(e))

@router.post("/register")
def register(request: RegisterRequest, auth_service: IAuthService = Depends(get_auth_service)):
    """
    Register a new user in the system.

    Args:
        request (RegisterRequest): Information required to create a new user.
        auth_service (IAuthService): Service for registering users.

    Returns:
        dict: Success message on successful registration.

    Raises:
        HTTPException: 400 on registration failure (e.g., user already exists).
    """
    try:
        # Attempt to register the new user
        auth_service.register(request)
        return {"message": "User registered successfully"}
    except Exception as e:
        # Registration failed (user exists, validation error, etc.)
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/refresh", response_model=LoginResponse)
def refresh_token(request: RefreshTokenRequest, auth_service: IAuthService = Depends(get_auth_service)):
    """
    Refresh an access token using a valid refresh token.

    Args:
        request (RefreshTokenRequest): Contains the refresh token.
        auth_service (IAuthService): Service for refreshing tokens.

    Returns:
        LoginResponse: New JWT access/refresh token pair.

    Raises:
        HTTPException: 401 on invalid or expired refresh token.
    """
    try:
        # Attempt to refresh the JWT tokens using the provided refresh token
        return auth_service.refresh_token(request.refresh_token)
    except Exception as e:
        # Token refresh failed (invalid/expired token)
        raise HTTPException(status_code=401, detail=str(e))

File: dotnet-backend/src/Api/Controllers/AuthController.cs
using Application.DTOs;
using Application.Interfaces;
using Microsoft.AspNetCore.Mvc;

namespace Api.Controllers;

/// <summary>
/// Controller for handling user authentication, registration, and token refresh operations.
/// </summary>
[ApiController]
[Route("api/[controller]")]
public class AuthController(IAuthService authService) : ControllerBase
{
    /// <summary>
    /// Authenticates the user and returns a JWT token if successful.
    /// </summary>
    /// <param name="request">The login credentials (username/email and password).</param>
    /// <returns>
    /// 200 OK with JWT token and user details on successful authentication;<br/>
    /// 401 Unauthorized if credentials are invalid.
    /// </returns>
    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        try
        {
            // Call the service to perform login and generate tokens
            var response = await authService.LoginAsync(request);
            return Ok(response);
        }
        catch (UnauthorizedAccessException)
        {
            // Return 401 if authentication fails
            return Unauthorized();
        }
    }

    /// <summary>
    /// Registers a new user with the provided registration information.
    /// </summary>
    /// <param name="request">User registration details.</param>
    /// <returns>
    /// 200 OK if registration is successful;<br/>
    /// 400 Bad Request if registration fails due to invalid data or duplicate accounts.
    /// </returns>
    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        try
        {
            // Register the user via the service layer
            await authService.RegisterAsync(request);
            return Ok();
        }
        catch (InvalidOperationException ex)
        {
            // Return 400 Bad Request with error message if registration fails
            return BadRequest(ex.Message);
        }
    }

    /// <summary>
    /// Refreshes the JWT token using a valid refresh token.
    /// </summary>
    /// <param name="request">The refresh token payload.</param>
    /// <returns>
    /// 200 OK with a new JWT token if refresh token is valid;<br/>
    /// 401 Unauthorized if refresh token is invalid or expired.
    /// </returns>
    [HttpPost("refresh")]
    public async Task<IActionResult> Refresh([FromBody] RefreshTokenRequest request)
    {
        try
        {
            // Request a new JWT using the refresh token
            var response = await authService.RefreshTokenAsync(request.RefreshToken);
            return Ok(response);
        }
        catch (UnauthorizedAccessException)
        {
            // Return 401 if refresh token is invalid or not found
            return Unauthorized();
        }
    }
}

/// <summary>
/// Represents a request to refresh a JWT using a refresh token.
/// </summary>
public record RefreshTokenRequest(string RefreshToken);
